File: messagematch/src/main/java/org/forwoods/messagematch/generate/nodegenerators/constraints/ProvidedConstraint.java
package org.forwoods.messagematch.generate.nodegenerators.constraints;

import java.math.BigDecimal;
import java.math.BigInteger;

public class ProvidedConstraint implements Constraint{

    private final Object value;
    public ProvidedConstraint(String val) {
        Object valStore;
        try {
            valStore = new BigInteger(val);
        }
        catch (NumberFormatException e) {
            try {
                valStore = new BigDecimal(val);
            }
            catch (NumberFormatException e2) {
                valStore = val;
            }
        }
        value = valStore;
    }

    @Override
    public boolean matches(Object o) {
        return o.equals(value);
    }

    @Override
    public Object generate() {
        return value;
    }
}

File: messagematch-typescript/src/generate/constraints/ProvidedConstraint.ts
import { Constraint } from '../ValueProvider';

export class ProvidedConstraint implements Constraint {
    private readonly value: bigint | number | string;

    constructor(val: string) {
        let valStore: bigint | number | string;
        try {
            // Try BigInt first (equivalent to Java's BigInteger)
            valStore = BigInt(val);
        } catch {
            try {
                // Then try number (equivalent to Java's BigDecimal)
                const num = Number(val);
                if (isNaN(num)) throw new Error();
                valStore = num;
            } catch {
                // Fall back to string
                valStore = val;
            }
        }
        this.value = valStore;
    }

    matches(o: any): boolean {
        return o === this.value || o?.toString() === this.value?.toString();
    }

    generate(): any {
        return this.value;
    }
}